lib/dispatcher/balanced-pool.js

```diff
@@ -146,6 +154,50 @@ class BalancedPool extends PoolBase {
       .map((p) => p[kUrl].origin)
   }
 
+  get [kFree] () {
+    return this[kClients].map(pool => pool[kFree]).reduce((a, b) => a + b, 0)
+  }
+
+  get [kConnected] () {
+    return this[kClients].map(pool => pool[kConnected]).reduce((a, b) => a + b, 0)
+  }
+
+  [kQueueResume] () {
+    if (!this[kMaxTotalConnections] || this[kQueued] === 0) return
+
+    // process an item from the queue if there are free connections
+    if (this[kConnected] < this[kMaxTotalConnections] || this[kFree] > 0) {
+      const item = this[kQueue].shift()
+      if (item) {
+        this[kQueued]--
+
+        // rather than loop synchronously, run one queue item per event loop
+        // since kConnected and kFree values change during async operations
+        if (this[kQueued] > 0) setTimeout(() => this[kQueueResume](), 0)
+      }
+
+      const { opts, handler } = item
+      const pool = this[kClients].find(pool => pool[kUrl].origin === opts.origin && pool[kFree] > 0)
+      if (pool) {
+        pool.dispatch(opts, handler)
+      } else {
+        this[kQueue].push({ opts, handler })
+        this[kQueued]++
+      }
+    }
+  }
+
+  [kDispatch] (opts, handler) {
+    // wrap handler.onComplete so it triggers kQueueResume after each request completes
+    const { onComplete } = handler
+    handler.onComplete = (...args) => {
+      setTimeout(this[kQueueResume](), 0)
+      if (onComplete) onComplete(...args)
```

Comment:
I mentioned this in https://github.com/nodejs/undici/issues/3268, but to put it in context: wrapping a handler's `onComplete` function appears to happen every time a request is attempted, so if a request is taken from the queue `x` times before finally being sent, this call to `kQueueResume` happens `x` times as well. Testing this locally I noted that it caused `kQueueResume` calls to increase exponentially over time, which is not at all ideal.

I tried adding a private symbol `handler[kResumable] = true` when it had been wrapped, to avoid re-wrapping, but it seems that `handler` is a brand new object every time `kDispatch` is called.